File: apps/api/app/db/models/message.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, Column, DateTime, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import relationship

from app.db.base import Base, TenantMixin


class Thread(Base, TenantMixin):
    __tablename__ = "threads"

    type = Column(String(20), nullable=False, default="direct")  # direct|group|announcement
    subject = Column(String(500), nullable=True)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)

    participants = relationship("ThreadParticipant", back_populates="thread", cascade="all, delete-orphan")
    messages = relationship("Message", back_populates="thread", cascade="all, delete-orphan",
                            order_by="Message.created_at")


class ThreadParticipant(Base):
    __tablename__ = "thread_participants"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    thread_id = Column(UUID(as_uuid=True), ForeignKey("threads.id"), nullable=False)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    role = Column(String(20), default="recipient")  # sender|recipient|cc
    read_at = Column(DateTime, nullable=True)
    archived = Column(Boolean, default=False)

    thread = relationship("Thread", back_populates="participants")


class Message(Base):
    __tablename__ = "messages"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    thread_id = Column(UUID(as_uuid=True), ForeignKey("threads.id"), nullable=False)
    sender_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    body = Column(Text, nullable=False)
    attachments = Column(JSONB, default=list)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    edited_at = Column(DateTime, nullable=True)

    thread = relationship("Thread", back_populates="messages")

File: packages/shared/src/types/message.ts
import type { BaseEntity, TenantScoped, ThreadType, ParticipantRole } from './common';

export interface Attachment {
  name: string;
  url: string;
  size: number;
  mime: string;
}

export interface Thread extends BaseEntity, TenantScoped {
  type: ThreadType;
  subject: string | null;
  created_by: string;
}

export interface ThreadParticipant {
  id: string;
  thread_id: string;
  user_id: string;
  role: ParticipantRole;
  read_at: string | null;
  archived: boolean;
}

export interface Message extends BaseEntity {
  thread_id: string;
  sender_id: string;
  body: string;
  attachments: Attachment[];
  edited_at: string | null;
}
